### PR TITLE
Add daemon mode flag

### PR DESCRIPTION
**Pull request by faisal-memon**
- Add a `daemon_mode` configurable to toggle running spiffe-helper as a daemon. When daemon mode is disabled x509 and/or jwts are downloaded and spiffe-helper exits.
- Add a `daemon-mode` cli flag to override config file setting and disable daemon mode.

### CODE REVIEW

File: cmd/spiffe-helper/main.go [hunk 1]
```diff
@@ -13,39 +13,43 @@ import (
 )
 
 func main() {
-	// 0. Load configuration
-	// 1. Create Sidecar
-	// 2. Run Sidecar's Daemon
-
 	configFile := flag.String("config", "helper.conf", "<configFile> Configuration file path")
-	exitWhenReady := flag.Bool("exitWhenReady", false, "Exit once the requested objects are retrieved")
+	disableDaemonMode := flag.Bool("disable_daemon_mode", false, "Exit once the requested objects are retrieved")
```
Comment (keeganwitt) [inline cmd/spiffe-helper/main.go:17]: Isn't the config `daemon_mode`, not `disable_daemon_mode`?

Comment (faisal-memon) [reply cmd/spiffe-helper/main.go:17]: Changed to daemon mode

File: cmd/spiffe-helper/config/config.go [hunk 2]
```diff
@@ -46,28 +48,36 @@ type JWTConfig struct {
 	JWTSVIDFilename string `hcl:"jwt_svid_file_name"`
 }
 
+func ParseFlags() (string, bool) {
+	configFile := flag.String("config", "helper.conf", "<configFile> Configuration file path")
+	daemonModeFlag := flag.Bool(daemonModeFlagName, true, "Exit once the requested objects are retrieved")
+	flag.Parse()
```
Comment (MarcosDY) [inline cmd/spiffe-helper/config/config.go:54]: why to parse here? generally it is expected than flags are parsed in main so it is easy to found,
is it for testing?

File: cmd/spiffe-helper/config/config.go [hunk 3]
```diff
@@ -140,16 +150,24 @@ func ValidateConfig(c *Config, exitWhenReady bool, log logrus.FieldLogger) error
...
-	x509EmptyCount := countEmpty(c.SVIDFileName, c.SVIDBundleFileName, c.SVIDKeyFileName)
-	jwtBundleEmptyCount := countEmpty(c.SVIDBundleFileName)
-	if x509EmptyCount == 3 && len(c.JWTSVIDs) == 0 && jwtBundleEmptyCount == 1 {
+	if !x509Enabled && !jwtBundleEnabled && !jwtSVIDsEnabled {
 		return errors.New("at least one of the sets ('svid_file_name', 'svid_key_file_name', 'svid_bundle_file_name'), 'jwt_svids', or 'jwt_bundle_file_name' must be fully specified")
 	}
 
-	if x509EmptyCount != 0 && x509EmptyCount != 3 {
-		return errors.New("all or none of 'svid_file_name', 'svid_key_file_name', 'svid_bundle_file_name' must be specified")
+	if isFlagPassed(daemonModeFlagName) {
```
Comment (MarcosDY) [inline cmd/spiffe-helper/config/config.go:164]: can you unify daemonModeFlag to a single value?
it must be calculated before getting here, so we can just set the value we get

File: cmd/spiffe-helper/main.go [hunk 4]
```diff
@@ -13,39 +12,41 @@ import (
...
 	defer stop()
 
-	hclConfig, err := config.ParseConfig(configPath)
+	configFile, daemonModeFlag := config.ParseFlags()
+	log.Infof("Using configuration file: %q", configFile)
+
+	hclConfig, err := config.ParseConfig(configFile)
 	if err != nil {
-		return fmt.Errorf("failed to parse %q: %w", configPath, err)
+		return fmt.Errorf("Failed to parse %q: %w", configFile, err)
```
Comment (MarcosDY) [inline cmd/spiffe-helper/main.go:35]:
```suggestion
		return fmt.Errorf("failed to parse %q: %w", configFile, err)
```

File: cmd/spiffe-helper/main.go [hunk 5]
```diff
@@ -13,39 +12,41 @@ import (
...
 	}
-	if err := config.ValidateConfig(hclConfig, exitWhenReady, log); err != nil {
-		return fmt.Errorf("invalid configuration: %w", err)
+
+	if err := config.ValidateConfig(hclConfig, daemonModeFlag, log); err != nil {
+		return fmt.Errorf("Invalid configuration: %w", err)
```
Comment (MarcosDY) [inline cmd/spiffe-helper/main.go:39]:
```suggestion
		return fmt.Errorf("invalid configuration: %w", err)
```

File: cmd/spiffe-helper/config/config.go (hunk 2 above)
Comment (faisal-memon) [reply cmd/spiffe-helper/config/config.go:54]: Yes for unit testing. See related tested here: https://github.com/spiffe/spiffe-helper/pull/161/files#diff-a7f12ddbf036810a09ab419309d276dae4b679371fc962c1ad5a63cd0cf790ccR400

Comment (faisal-memon) [reply cmd/spiffe-helper/config/config.go:54]: i can move back to main and have duplicate for tests if you think thats better.

File: cmd/spiffe-helper/config/config.go (hunk 3 above)
Comment (faisal-memon) [reply cmd/spiffe-helper/config/config.go:164]: Can move to `ParseConfig()`. Is that what you're thinking?

File: cmd/spiffe-helper/config/config.go [hunk 6]
```diff
@@ -140,16 +150,24 @@ func ValidateConfig(c *Config, exitWhenReady bool, log logrus.FieldLogger) error
...
+		// If daemon mode is set by CLI this takes precedence
+		c.DaemonMode = &daemonModeFlag
+	} else if c.DaemonMode == nil {
+		// If daemon mode is not set, then default to true
+		daemonMode := true
```
Comment (keeganwitt) [inline cmd/spiffe-helper/config/config.go:169]: Should the default be false? This is changing default behavior.

File: pkg/sidecar/config.go [hunk 7]
```diff
@@ -5,7 +5,6 @@ import (
 )
 
 type Config struct {
-	// If true, merge intermediate certificates into Bundle file instead of SVID file.
```
Comment (keeganwitt) [inline pkg/sidecar/config.go:8]: How is this change related? Was it's removal intentional?

Comment (faisal-memon) [reply pkg/sidecar/config.go:8]: accident

**Comment by keeganwitt** [comment]
I'm testing this code now. Somehow the flag gets parsed to false, even when true is passed in as an argument. I'm not sure why yet.

File: cmd/spiffe-helper/config/config.go [hunk 8]
```diff
@@ -46,28 +48,36 @@ type JWTConfig struct {
 	JWTSVIDFilename string `hcl:"jwt_svid_file_name"`
 }
 
+func ParseFlags() (string, bool) {
+	configFile := flag.String("config", "helper.conf", "<configFile> Configuration file path")
+	daemonModeFlag := flag.Bool(daemonModeFlagName, true, "Exit once the requested objects are retrieved")
```
Comment (keeganwitt) [inline cmd/spiffe-helper/config/config.go:53]: Usually the doc for a flag describes what it does if true I think. So this is worded backwards from what one might expect.

File: cmd/spiffe-helper/config/config.go (hunk 2 above)
Comment (keeganwitt) [inline cmd/spiffe-helper/config/config.go:54]: For some reason, in my testing adding
```go
logrus.Infof("Parsed daemon flag of %t", *daemonModeFlag)
```
after this line logs true, even if I passed `false`, `FALSE`, or `0` on the CLI.

File: cmd/spiffe-helper/config/config.go (hunk 8 above)
Comment (faisal-memon) [reply cmd/spiffe-helper/config/config.go]: updated description.

File: cmd/spiffe-helper/config/config.go (hunk 2 above)
Comment (MarcosDY) [reply cmd/spiffe-helper/config/config.go]: that will be great

File: cmd/spiffe-helper/config/config.go (hunk 3 above)
Comment (MarcosDY) [reply cmd/spiffe-helper/config/config.go:164]: sure

File: cmd/spiffe-helper/config/config.go (hunk 2 above)
Comment (faisal-memon) [reply cmd/spiffe-helper/config/config.go]: done

File: cmd/spiffe-helper/config/config.go (hunk 3 above)
Comment (faisal-memon) [reply cmd/spiffe-helper/config/config.go]: Moved that logic to a separate function

File: pkg/sidecar/config.go (hunk 7 above)
Comment (faisal-memon) [reply pkg/sidecar/config.go]: added back in

**Review by MarcosDY — APPROVED** [review]
